### DISCUSSION TITLE
Error for C code in Moose

### DISCUSSION THREAD

**Discussion by caoductoan379** [idaholab/moose#16357] [Q&A General] [answered]
Dear All for help

I modified the code as follows: 

 #include "PFFracBulkRateMaterial.h"
#include "Function.h"

registerMooseObject("PhaseFieldApp", PFFracBulkRateMaterial);

template<>
InputParameters validParams<PFFracBulkRateMaterial>()
{
  InputParameters params = validParams<Material>();
  params.addClassDescription("Material properties used in phase-field fracture damage evolution kernel");
  params.addParam<FunctionName>("function", "Function describing energy release rate type parameter distribution");
  params.addParam<Real>("gc", 1.0, "Energy release rate type parameter");

  return params;
}

PFFracBulkRateMaterial::PFFracBulkRateMaterial(const InputParameters & parameters) :
    Material(parameters),
    _gc(getParam<Real>("gc")),
    _gc_prop(declareProperty<Real>("gc_prop")),
    _function_prop(isParamValid("function") ? &getFunction("function") : NULL);
{
}

void
PFFracBulkRateMaterial::initQpStatefulProperties()
{
  _gc_prop[_qp] = _gc;
}

void
PFFracBulkRateMaterial::computeQpProperties()
{
  _gc_prop[_qp] = _gc;
  /**
   * This function computes heterogeneous gc
   * User should override this function if heterogenities needs consideration
   */
  getProp();
}

void
PFFracBulkRateMaterial::getProp()
{
  if (_function_prop != NULL)
    _gc_prop[_qp] = _function_prop->value(_t, _q_point[_qp]);
}

#include "Material.h"
#include "Function.h"

/**
 * Phase-field fracture
 * This class obtains critical energy release rate (gc) value
 * Used by PFFRacBulkRate
 */

class PFFracBulkRateMaterial;

template<>
InputParameters validParams<PFFracBulkRateMaterial>();

class PFFracBulkRateMaterial : public Material
{
public:
  PFFracBulkRateMaterial(const InputParameters & parameters);

protected:
  virtual void initQpStatefulProperties();
  virtual void computeQpProperties();
  /**
   * This function obtains the value of gc
   * Must be overidden by the user for heterogeneous gc
   */
  virtual void getProp();

  ///Input parameter for homogeneous gc
  Real _gc;

  ///Material property where the gc values are stored
  MaterialProperty<Real> &_gc_prop;
  ///Function to specify varying gc
  Function * _function_prop;

private:

};

I compiled my code in phase field model. I received the error as follows: 

/projects/moose/modules/phase_field/src/materials/PFFracBulkRateMaterial.C: In constructor 'PFFracBulkRateMaterial::PFFracBulkRateMaterial(const InputParameters&)':
/projects/moose/modules/phase_field/src/materials/PFFracBulkRateMaterial.C:26:45: error: invalid conversion from 'const Function*' to 'Function*' [-fpermissive]
     _function_prop(isParamValid("function") ? &getFunction("function") : NULL);
                                             ^
/projects/moose/modules/phase_field/src/materials/PFFracBulkRateMaterial.C:26:79: error: expected '{' before ';' token
     _function_prop(isParamValid("function") ? &getFunction("function") : NULL);
                                                                               ^
/projects/moose/modules/phase_field/src/materials/PFFracBulkRateMaterial.C: At global scope:
/projects/moose/modules/phase_field/src/materials/PFFracBulkRateMaterial.C:27:1: error: expected unqualified-id before '{' token
 {
 ^
/projects/moose/framework/build.mk:118: recipe for target '/projects/moose/modules/phase_field/build/unity_src/materials_Unity.x86_64-pc-linux-gnu.opt.lo' failed
make: *** [/projects/moose/modules/phase_field/build/unity_src/materials_Unity.x86_64-pc-linux-gnu.opt.lo] Error 1

Could everyone help to correct this error ? i really appreciate it .

Kind regards, 

Toan

**Comment by jessecarterMOOSE** [comment, marked as answer]
addParam needs a type: ie.
`params.addParam<FunctionName>("function", "Function describing energy release rate type parameter distribution");`

Also no semicolon is needed after the constructor
`_function_prop(isParamValid("function") ? &getFunction("function") : NULL)`

**Reply by jessecarterMOOSE** [reply]
Same goes for your `gc` param. Perhaps `params.addParam<Real>`?

Also I think when you declare the property, that is going to need a type: `declareProperty<Real>`

**Reply by caoductoan379** [reply]
Thanks  Jessecarter. I will try it and i will let you know what happens. 

Kind regards, 

Toan

**Reply by caoductoan379** [reply]
Thanks  Jessecarter.  it is working. 

Thank for your help. 

Kind regards, 

Toan
